app/routes/devices.$id.hardware.$hardware_name.calibrate.tsx

```diff
@@ -0,0 +1,252 @@
+    case Intent.Enum.start_calibration_procedure:
+      try {
+        await Evolver.startCalibrationProcedureHardwareHardwareNameCalibratorProcedureStartPost(
+          {
+            path: {
+              hardware_name: submission.value.hardware_name,
+            },
+            client: evolverClient,
+          },
+        );
+      } catch (error) {
+        return submission.reply({
+          formErrors: ["unable to start calibration"],
+        });
+      }
+      break;
+    default:
+      return submission.reply();
+  }
+  return null;
+}
+
+export async function loader({ params }: LoaderFunctionArgs) {
+  const { id, hardware_name } = params;
+  const targetDevice = await db.device.findUnique({ where: { device_id: id } });
+  if (!targetDevice) {
+    return json({ actions: [] });
+  }
+  const { url } = targetDevice;
+  const evolverClient = createClient({
+    baseUrl: url,
+  });
+  const { data } =
+    await Evolver.getCalibratorActionsHardwareHardwareNameCalibratorProcedureActionsGet(
+      {
+        path: {
+          hardware_name: hardware_name ?? "",
+        },
+        client: evolverClient,
+      },
+    );
+
+  return json({ actions: data?.actions });
+}
+
+const CalibrationProcedure = ({ actions }) => {
+  const submit = useSubmit();
+  const { id, hardware_name } = useParams();
+
+  return actions.map((action, ix) => {
+    const dispatchAction = (actionFormData: object) => {
+      const formData = new FormData();
+      formData.append("id", id ?? "");
+      formData.append("intent", Intent.Enum.dispatch_action);
+      formData.append("hardware_name", hardware_name ?? "");
+      formData.append("action_name", action.name);
+      formData.append("payload", JSON.stringify(actionFormData));
+      submit(formData, {
+        method: "POST",
+      });
+    };
+    return (
+      <ClientOnly
+        key={action.description}
+        fallback={<span className="skeleton h-32"></span>}
+      >
+        {() => (
+          <CalibratorActionForm
+            action={action}
+            index={ix}
+            dispatchAction={dispatchAction}
+          />
+        )}
+      </ClientOnly>
+    );
+  });
+};
+
+export default function CalibrateHardware() {
+  const { actions } = useLoaderData<typeof loader>();
+  const { id, hardware_name } = useParams();
+  const hasActions = actions && actions.length > 0;
+  const calibrationButtonCopy = hasActions ? "restart" : "start";
+
+  const actionData = useActionData<typeof action>();
+  const submit = useSubmit();
+
+  useEffect(() => {
+    if (actionData?.error) {
+      if (typeof actionData.error === "string") {
+        notify.error(actionData.error);
+      }
+      if (typeof actionData.error === "object") {
+        const errorMessages: string[] = [];
+        Object.entries(actionData.error).forEach(([key, value]) => {
+          errorMessages.push(`${key}: ${value}`);
+        });
+        errorMessages.forEach((message) => {
+          notify.error(message);
```

Comment:
Kind of looks like the `errorMessage` variable is unnecessary, could the messages be directly notified?
```ts
Object.entries(actionData.error).forEach(([key, value]) => {
  notify.error(`${key}: ${value}`);
}
```

Reply:
yeah makes sense, will fix.